src/core/datastore/deltaset.js

```diff
@@ -0,0 +1,159 @@
+      return offlineRepo.read(queryCacheCollectionName, queryCacheQuery)
+        .then((deltaSetQueryDocs = []) => {
+          if (deltaSetQueryDocs.length > 0) {
+            return deltaSetQueryDocs[0];
+          }
+
+          return {
+            _id: generateEntityId(),
+            collectionName: collectionNameWithTag,
+            query: serializedQuery
+          };
+        });
+    });
+}
+
+function updateDeltaSetQuery(deltaSetQuery, response) {
+  return repositoryProvider.getOfflineRepository()
+    .then((offlineRepo) => {
+      deltaSetQuery.lastRequest = response.headers.get(xKivneyRequestStartHeader);
+      return offlineRepo.update(queryCacheCollectionName, deltaSetQuery);
+    });
+}
+
+function readDocs(collectionNameWithTag, query) {
+  return repositoryProvider.getOfflineRepository()
+    .then((offlineRepo) => {
+      return offlineRepo.read(collectionNameWithTag, query);
+    });
+}
+
+function deleteDocs(collectionNameWithTag, deleted = []) {
+  return repositoryProvider.getOfflineRepository()
+    .then((offlineRepo) => {
+      if (isArray(deleted) && deleted.length > 0) {
+        const deletedIds = deleted.map((doc) => doc._id);
+        const deleteQuery = new Query().contains('_id', deletedIds);
+        return offlineRepo.delete(collectionNameWithTag, deleteQuery)
+      }
+
+      return deleted;
+    });
+}
+
+function updateDocs(collectionNameWithTag, changed = []) {
+  return repositoryProvider.getOfflineRepository()
+    .then((offlineRepo) => {
+      if (isArray(changed) && changed.length > 0) {
+        return offlineRepo.update(collectionNameWithTag, changed)
+      }
+
+      return changed;
+    });
+}
+
+function makeDeltaSetRequest(collectionNameWithTag, deltaSetQuery, query, options) {
+  const collectionName = stripTagFromCollectionName(collectionNameWithTag);
+  const client = Client.sharedInstance();
+  const request = new KinveyRequest({
+    authType: AuthType.Default,
+    method: RequestMethod.GET,
+    url: format({
+      protocol: client.apiProtocol,
+      host: client.apiHost,
+      pathname: buildCollectionUrl(collectionName, '_deltaset'),
+      query: { since: deltaSetQuery.lastRequest }
+    }),
+    query,
+    timeout: options.timeout,
+    followRedirect: options.followRedirect,
+    cache: options.cache,
+    properties: options.properties,
+    skipBL: options.skipBL,
+    trace: options.trace,
+    client
+  });
+
+  return request.execute()
+    .then((response) => {
+      return updateDeltaSetQuery(deltaSetQuery, response).then(() => response.data);
+    })
+    .then((data) => {
+      return deleteDocs(collectionNameWithTag, data.deleted)
+        .then(() => updateDocs(collectionNameWithTag, data.changed))
+        .then(() => data);
+    });
+}
+
+function makeRequest(collectionNameWithTag, deltaSetQuery, query, options) {
+  if (deltaSetQuery && deltaSetQuery.lastRequest) {
+    if (query && (query.skip != null || query.limit != null)) {
```

Comment:
@georgiwe re: your second point - isn't the idea that `limit` and `skip` together with a deltaset request is not supported? If I'm reading this code snippet correctly, making a `deltaSetQuery` with `limit` and `skip` is not supported. I think that's reasonable. Thoughts?

Reply:
Sure, and I didn't think of this at the time, but yesterday I kinda changed my mind. And this may affect the way we treat skip and limit, if we want to keep some consistency.

It seems important to keep `.find()`s fully functional. In all store types, with all settings. If not, user would need to write their app in a less convenient way and keep track of what works when. If we start throwing an error in CacheStore.find() when skip and limit is passed, and deltaset is enabled, as a developer, it would be difficult to write my app. I may want to make some skip and limit requests - should I disable delta set for those requests only?

One way to keep them, is not send them as part of the delta set request (ignore them), but apply them to the request to the offline store, [here](https://github.com/Kinvey/js-sdk/blob/MLIBZ-2316_server_side_delta_set/src/core/datastore/deltaset.js#L152). This way we "update the local store", and then read whatever the user is trying to read, from a "synced" place.

This approach may introduce some confusion as it's different from `pull()`, but I think it isn't too bad.

Does this make sense, or am I missing something?

Reply:
This makes sense. I think your suggestion also makes sense. @tejasranade thoughts?

Reply:
Yes, this is what I'm suggesting. I don't think it conflicts. I think skip and limit help you target a specific subset of data (just like you said), usually when paging things, but not always. Not just from the server, but from any data source. Cause if you had to pull everything, just to take the last 3 entities, things don't make sense, regardless of the data source.

I guess turning off deltaset would make the API easier to work with, but I think what I suggested is better - we get the best of both worlds. We get performance, we get the expected result and an easy to use API. What do you dislike about it - you feel it's semantically wrong?

Ultimately, I'm fine with whatever you decide - the most important part is to have an easy to use API, which works correctly and as expected.